src/transformStateWithClones.js

```diff
@@ -7,7 +7,39 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [state];
+
+  actions.forEach((action) => {
+    let newState;
+
+    const prevState = { ...stateHistory[stateHistory.length - 1] };
+
+    switch (action.type) {
+      case 'clear':
+        newState = {};
+        break;
+
+      case 'addProperties':
+        newState = { ...prevState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        newState = { ...prevState };
+
+        action.keysToRemove.forEach((key) => {
+          delete newState[key];
+        });
+        break;
+
+      default:
+        newState = prevState;
+        break;
+    }
+    stateHistory.push(newState);
+  });
+  stateHistory.shift();
```

Comment:
The use of `shift` at line 40 is unnecessary and potentially inefficient since it modifies the original array. Instead, you can initialize `stateHistory` as an empty array and then push state copies into it as you process actions. This way, you won't need to remove the initial state at the end.